### PR TITLE
Catalog update

### PR DESCRIPTION
**Pull request by geomacy**
For AMP-982, work in progress.

Updates BrooklynObject and BrooklynObjectSpec to support nested catalogItemIds.
### TODO:
- doesn't include changes to persistence to allow the nested item ids to be persisted and retrieved during rebind
- Need to review what changes may be needed to EntityTransientCopyInternal.
- Looks like catalogItemIdIfNotNull can be removed.
### Things to review

Everything, really; but some things I made a note of particularly
- review places where setCatalogItemIdIfNotNull has been replaced with nestCatalogItemIds and see if they make sense from a sanity viewpoint
- review use of setCatalogItemIds()

### CODE REVIEW

**Comment by geomacy** [comment]
still no good sorry

**Comment by geomacy** [comment]
ah got it the commits above are in reverse chronological order; my changes are the last four, as you see.

File: core/src/main/java/org/apache/brooklyn/core/catalog/internal/CatalogUtils.java [hunk 1]
```diff
@@ -170,6 +173,13 @@ public static void installLibraries(ManagementContext managementContext, @Nullab
     public static String getCatalogItemIdFromLoader(BrooklynClassLoadingContext loader) {
         if (loader instanceof OsgiBrooklynClassLoadingContext) {
             return ((OsgiBrooklynClassLoadingContext)loader).getCatalogItemId();
+        } else if (loader instanceof BrooklynClassLoadingContextSequential) {
+            final Iterator<BrooklynClassLoadingContext> iterator = ((BrooklynClassLoadingContextSequential) loader).getPrimaries().iterator();
+            if (iterator.hasNext()) {
+                BrooklynClassLoadingContext osgiLoader = iterator.next();
+                return getCatalogItemIdFromLoader(osgiLoader);
+            }
+            else return null;
```
Comment (neykov) [inline core/src/main/java/org/apache/brooklyn/core/catalog/internal/CatalogUtils.java:182]: Need to revisit this logic in light of latest updates.

File: core/src/main/java/org/apache/brooklyn/core/catalog/internal/CatalogItemDtoAbstract.java [hunk 2]
```diff
@@ -45,6 +45,8 @@
 import com.google.common.collect.Iterables;
 import com.google.common.collect.Sets;
 
+// TODO add support for nested catalog items, implement nestCatalogItemId in terms of symbolicName/Version
+// TODO also getCatalogItemSuperIds.
```
Comment (neykov) [inline core/src/main/java/org/apache/brooklyn/core/catalog/internal/CatalogItemDtoAbstract.java:49]: I think we should either override with NOPs, or just keep the default behaviour from `AbstractBrooklynObject`. The catalog items don't know what's the meaning of the plan in them.

**Review by neykov — APPROVED** [review]
Looks good so far.

**Comment by geomacy** [comment]
Retiring this PR in favour of https://github.com/apache/brooklyn-server/pull/338

**Pull request closed without merging**